De-duplicate Webpack build, especially `components/`

nylen opened the pull request:
While looking at #926 I noticed that we are probably duplicating some code in the `components/` directory across built files.  For example, we import `IconButton` in both [`editor/header/tools/preview-button.js`](https://github.com/WordPress/gutenberg/blob/739ae9808932b02c92e770033a1c2f7e2db4c5dc/editor/header/tools/preview-button.js#L9) and [`blocks/editable/format-toolbar.js`](https://github.com/WordPress/gutenberg/blob/d3c4425d4fa15ff4781f46de5fb925767e2b0d5e/blocks/editable/format-toolbar.js#L4).

If we split `components/` out into a separate build + entry point, we should be able to resolve this issue, and also make our common components available via `wp.components`, which seems good for reusability.

This PR is an attempt at doing this, but it's not working for some reason.  Here are the build sizes before this PR:

```
$ ls -la */build/*.js
271232 blocks/build/index.js
  3119 date/build/index.js
466975 editor/build/index.js
 14447 element/build/index.js
 17481 i18n/build/index.js
```

and after:

```
$ ls -la */build/*.js
271230 blocks/build/index.js
118012 components/build/index.js
  3119 date/build/index.js
467031 editor/build/index.js
 14389 element/build/index.js
 17481 i18n/build/index.js
```

nylen committed:
Build `components/` directory separately

nylen committed:
Fix the build

`component` depends on `element`. :(

BE-Webdesign commented:
Looks good to me.  Will we also need to go through the files and change where modules are being imported from?  Or is webpack smart enough to not duplicate once a build is created for that directory?  Here is the current list to illustrate what I mean.

```
gutenberg\blocks\block-controls\index.js
9:import Toolbar from 'components/toolbar';

gutenberg\blocks\editable\format-toolbar.js
4:import IconButton from 'components/icon-button';
5:import Toolbar from 'components/toolbar';

gutenberg\blocks\editable\index.js
13:import Toolbar from 'components/toolbar';

gutenberg\blocks\library\button\index.js
4:import IconButton from 'components/icon-button';

gutenberg\blocks\library\embed\index.js
4:import Button from 'components/button';
5:import Placeholder from 'components/placeholder';

gutenberg\blocks\library\image\index.js
4:import Placeholder from 'components/placeholder';

gutenberg\blocks\media-upload-button\index.js
6:import Button from 'components/button';

gutenberg\components\panel\body.js
5:import IconButton from 'components/icon-button';

gutenberg\components\placeholder\index.js
10:import Dashicon from 'components/dashicon';

gutenberg\editor\block-mover\index.js
9:import IconButton from 'components/icon-button';

gutenberg\editor\block-switcher\index.js
11:import IconButton from 'components/icon-button';
12:import Dashicon from 'components/dashicon';

gutenberg\editor\header\mode-switcher\index.js
9:import Dashicon from 'components/dashicon';

gutenberg\editor\header\saved-state\index.js
10:import Dashicon from 'components/dashicon';

gutenberg\editor\header\tools\index.js
10:import IconButton from 'components/icon-button';

gutenberg\editor\header\tools\preview-button.js
9:import IconButton from 'components/icon-button';

gutenberg\editor\header\tools\publish-button.js
9:import Button from 'components/button';

gutenberg\editor\inserter\index.js
10:import IconButton from 'components/icon-button';

gutenberg\editor\inserter\menu.js
11:import Dashicon from 'components/dashicon';
12:import withFocusReturn from 'components/higher-order/with-focus-return';

gutenberg\editor\modes\visual-editor\block.js
14:import Toolbar from 'components/toolbar';

gutenberg\editor\sidebar\block-inspector\index.js
9:import Panel from 'components/panel';
10:import PanelHeader from 'components/panel/header';
11:import PanelBody from 'components/panel/body';

gutenberg\editor\sidebar\discussion-panel\index.js
11:import PanelBody from 'components/panel/body';
12:import FormToggle from 'components/form-toggle';

gutenberg\editor\sidebar\featured-image\index.js
11:import Button from 'components/button';
12:import PanelBody from 'components/panel/body';
14:import Spinner from 'components/spinner';

gutenberg\editor\sidebar\index.js
9:import withFocusReturn from 'components/higher-order/with-focus-return';

gutenberg\editor\sidebar\post-excerpt\index.js
10:import PanelBody from 'components/panel/body';

gutenberg\editor\sidebar\post-schedule\clock.js
5:import Button from 'components/button';

gutenberg\editor\sidebar\post-settings\index.js
10:import Panel from 'components/panel';
11:import PanelHeader from 'components/panel/header';
12:import IconButton from 'components/icon-button';

gutenberg\editor\sidebar\post-status\index.js
11:import PanelBody from 'components/panel/body';
12:import FormToggle from 'components/form-toggle';

gutenberg\editor\sidebar\post-trash\index.js
10:import Button from 'components/button';
11:import Dashicon from 'components/dashicon';
```

nylen commented:
> Or is webpack smart enough to not duplicate once a build is created for that directory?

This is the part that's not working yet - I would expect to see the 118kb file for `components/build/index.js`, and then I would expect to see slightly more than this 118kb (due to existing duplication) moved out of `blocks/build/index.js` and `editor/build/index.js`.

I suspect something is still missing in the webpack config.

nylen committed:
Enqueue wp.components script

nylen committed:
Consolidate entry points into one variable

nylen committed:
Register all `wp.` libraries as Webpack externals

nylen commented:
Before this PR there's actually a lot of duplication in the built files.  Multiple copies of `Dashicon`, for example.  The latest commits improve things somewhat - here are the new build sizes:

```
$ ls -la */build/*.js
253181 blocks/build/index.js
 96260 components/build/index.js
  3119 date/build/index.js
334667 editor/build/index.js
 14554 element/build/index.js
 17483 i18n/build/index.js
```

However, there is still a lot of duplication - `IconButton` and `Dashicon` are each present in the `blocks`, `editor`, and `components` builds.  To achieve full separation between our built files, we will have to make further changes:

```diff
- import IconButton from 'components/icon-button';
+ import { IconButton } from 'components';
```

nylen committed:
Restructure imports to top-level entry points only

nylen committed:
Fix duplicate imports

nylen committed:
Fix naming of `withFocusReturn` HoC

nylen committed:
Fix the tests

- Fix entry point order to match actual dependencies
- Fix externals (`wp.*` externals not needed)

nylen commented:
This PR should be ready for review now.  It shaves 200 KB off of the production build size by adding a new `components/build/index.js` output file and then eliminating a lot of duplication across the various files built by Webpack.

It works by registering each of our top-level entry points like `components` as an external library using the [Webpack `externals` option](https://webpack.js.org/configuration/externals/).  Behind the scenes, then, Webpack will transform `import { thing } from 'components'` into something that translates to `wp.components.thing`.  Previously, all imports from e.g. `components` were being included in each output file.

I saw https://github.com/WordPress/gutenberg/pull/901#issuecomment-304316457 after I already started restructuring imports in this way. The one difference is that everything should be imported from `components` directly, instead of subdirectories like `components/higher-order`.

Current sizes:

```
$ ls -la */build/*.js
179431 blocks/build/index.js
 96343 components/build/index.js
  3119 date/build/index.js
258059 editor/build/index.js
 14550 element/build/index.js
 17483 i18n/build/index.js
```

Final statistics:

```
773254  Build size on `master`
568985  Build size on this branch
204269  Savings (26% of original size)
```

nylen committed:
Fix `wp-components` stylesheet and dependencies on editor page

youknowriad commented on components/index.js:
```diff
@@ -0,0 +1,10 @@
+export { default as Spinner } from './spinner';
+export { default as Toolbar } from './toolbar';
+
+export { default as withFocusReturn } from './higher-order/with-focus-return';
```
I don't think we should export this in the root module. I'm thinking we should create a `index.js` in the `./higher-order/` folder to be able to import those as `import { withFocusReturn } from 'components/higher-order'`;

youknowriad replied:
Oh I just saw your comment https://github.com/WordPress/gutenberg/pull/929#issuecomment-304771336

I still think we should make a distinction between importing components and HoC.

nylen replied:
How do we make this work like `wp.components.something` behind the scenes?

I think this could be solved separately, in any case.

BE-Webdesign reviewed (approved):
Nice Work! Verified that the build size is significantly smaller. LGTM, any issues can be cycled back to in another PR I think as this resolves things for the current structure.

nylen commented:
I've created #941 to discuss next steps.

iseulde commented:
I think this broke master. There's an error if you open the inspector.

youknowriad commented:
Confirmed this broke a lot of components. I fixed this in a branch https://github.com/WordPress/gutenberg/pull/945/commits/17741f48b308c508e602854cb32a3e024df87534 but we should probably fix this separately
